# Telecom Customer Churn
The goal is to perform exploratory data analysis and understand how these attributes affect churn

Import required libraries

In [42]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

Read the data

In [43]:
df=pd.read_csv("telecom_retention_data.csv")

In [44]:
df.sample(3)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
1609  3584-WKTTW    Male              0     Yes         No      51   
5368  6168-WFVVF  Female              1      No         No       3   
2830  1699-UOTXU    Male              0      No         No      60   

     PhoneService MultipleLines InternetService OnlineSecurity  ...  \
1609          Yes           Yes     Fiber optic            Yes  ...   
5368          Yes            No     Fiber optic             No  ...   
2830          Yes            No             DSL            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
1609              Yes         Yes         Yes             Yes        One year   
5368               No          No          No              No  Month-to-month   
2830              Yes          No          No              No        Two year   

     PaperlessBilling     PaymentMethod MonthlyCharges  TotalCharges Churn  
1609              Yes  Electronic check         111.55       5720.35    No  
5368              Yes  Electronic check          70.30         235.5   Yes  
2830               No  Electronic check          61.40       3638.25    No  

[3 rows x 21 columns]

Analyse the data

In [45]:
df.shape

(7043, 21)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


Observations:
1. Customer churn data contains 7043 rows and 21 columns.
2. The columns present in data are customerID gender, SeniorCitizen, Partner, Dependents, tenure, PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaperlessBilling, PaymentMethod, MonthlyCharges, TotalCharges, Churn   

Correct the datatypes

In [47]:
df['TotalCharges']=pd.to_numeric(df['TotalCharges'],errors='coerce')

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


Check for null values

In [49]:
df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [50]:
df.dropna(inplace=True)

In [51]:
df.shape

(7032, 21)

Observation:
1. There are 11 null entries in the Totalcharges coloumn

In [52]:
numeric_feature=[feature for feature in df if df[feature].dtype!='object']
print(len(numeric_feature))
categorical_feature=[feature for feature in df if df[feature].dtype=='object']
print(len(categorical_feature))

4
17


Observation:
1. There are 4 categorical columns and 17 numerical columns present in our dataset

Check the distribution of data

In [53]:
c=df['Churn'].value_counts()
print(c)
ratio=c[1]/c[0]
print(ratio)

Churn
No     5163
Yes    1869
Name: count, dtype: int64
0.36199883788495063


Observation:
1. Data is imbalanced and number of churner to non churner ratio is 0.36